app/src/main/java/org/oppia/app/utility/Logger.kt

```diff
@@ -0,0 +1,197 @@
+    write(LogLevel.DEBUG, tag, msg, tr)
+  }
+
+  /**
+   *
+   * @param tag Used to identify the source of a log message.  It usually identifies
+   * the class or activity where the log call occurs.
+   * @param msg The message you would like logged.
+   */
+  fun i(tag: String, msg: String) {
+    write(LogLevel.INFO, tag, msg)
+  }
+
+  /**
+   *
+   * @param tag Used to identify the source of a log message.  It usually identifies
+   * the class or activity where the log call occurs.
+   * @param msg The message you would like logged.
+   * @param tr  An exception to log
+   */
+  fun i(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.INFO, tag, msg, tr)
+  }
+
+  /**
+   *
+   * @param tag Used to identify the source of a log message.  It usually identifies
+   * the class or activity where the log call occurs.
+   * @param msg The message you would like logged.
+   */
+  fun w(tag: String, msg: String) {
+    write(LogLevel.WARNING, tag, msg)
+  }
+
+  /**
+   *
+   * @param tag Used to identify the source of a log message.  It usually identifies
+   * the class or activity where the log call occurs.
+   * @param msg The message you would like logged.
+   * @param tr  An exception to log
+   */
+  fun w(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.WARNING, tag, msg, tr)
+  }
+
+  /**
+   *
+   * @param tag Used to identify the source of a log message.  It usually identifies
+   * the class or activity where the log call occurs.
+   * @param msg The message you would like logged.
+   */
+  fun e(tag: String, msg: String) {
+    write(LogLevel.ERROR, tag, msg)
+  }
+
+  /**
+   *
+   * @param tag Used to identify the source of a log message.  It usually identifies
+   * the class or activity where the log call occurs.
+   * @param msg The message you would like logged.
+   * @param tr  An exception to log
+   */
+  fun e(tag: String, msg: String, tr: Throwable) {
+    write(LogLevel.ERROR, tag, msg, tr)
+  }
+
+  private fun isLogEnable(logLevel: LogLevel): Boolean {
+    return GLOBAL_LOG_LEVEL.logLevel < logLevel.logLevel
+  }
+
+  private fun write(logLevel: LogLevel, tag: String, log: String) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, log)
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg = Calendar.getInstance().time.toString() + "\t" + logLevel.name + "/" + tag + ": " + log
+      write(msg)
+    }
+  }
+
+  private fun write(logLevel: LogLevel, tag: String, log: String, tr: Throwable) {
+    if (isLogEnable(logLevel) && ENABLE_CONSOLE_LOG) {
+      Log.println(logLevel.logLevel, tag, log + "\n" + Log.getStackTraceString(tr))
+    }
+    if (isLogEnable(logLevel) && ENABLE_FILE_LOG) {
+      val msg =
+        Calendar.getInstance().time.toString() + "\t" + logLevel.name + "/" + tag + ": " + log + "\n" + Log.getStackTraceString(
+          tr
+        )
+      write(msg)
+    }
+  }
+
+  private fun write(text: String) {
```

Comment:
We shouldn't perform I/O like this on the main thread, as it can be costly. I suggest instead injecting the ``@BlockingExecutor`` coroutine dispatcher and scheduling writes to that executor. This may give us some throughput issues in the long-run, but it'll guarantee no main thread slow-downs or strict mode violations, and the blocking executor ensures the logs are written in order.